subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -136,6 +169,28 @@ private SnapshotHierarchy removeUnwatchableFileSystems(SnapshotHierarchy root, I
         return invalidatedRoot;
     }
 
+    public SnapshotHierarchy removeUnwatchableContentOnBuildStart(SnapshotHierarchy root, Invalidator invalidator) {
+        SnapshotHierarchy newRoot = root;
+        newRoot = removeUnprovenHierarchies(invalidator, newRoot);
+        watchProbesByHierarchy.clear();
+        watchProbesByPath.clear();
+        return newRoot;
+    }
+
+    @Nonnull
+    private SnapshotHierarchy removeUnprovenHierarchies(Invalidator invalidator, SnapshotHierarchy root) {
+        SnapshotHierarchy newRoot = root;
+        for (WatchProbe watchProbe : watchProbesByHierarchy.values()) {
+            if (watchProbe.leftArmed()) {
+                if (recentlyUsedHierarchies.remove(watchProbe.watchableHierarchy)) {
+                    LOGGER.warn("Invalidating hierarchy because watch probe hasn't been triggered {}", watchProbe.watchableHierarchy);
```

Comment:
We need to check in the FS-watching tests that we don't print out this message. I am pretty sure that no watch probe triggers on Linux currently.

Reply:
I added a check.